client/internal/bootstrap/auth_test.go

```diff
@@ -16,188 +17,159 @@ import (
-					TenantID:     testTenantID,
-				}
-
-				token, err := bootstrapClient.getAccessToken("", testResource, cloudProviderConfig)
-				Expect(err).To(BeNil())
-				Expect(token).To(Equal("token"))
-			})
-		})
-
-		When("service principal client secret contains certificate data", func() {
-			It("should return a corresponding SPN access token using certificate auth", func() {
+				config.CloudName = azure.PublicCloud.Name
+				config.ClientID = "service-principal-id"
+				config.ClientSecret = "certificate:" + certData
+			},
+			expectToken: "",
+			expectErr:   errors.New("generating SPN access token with certificate"),
+		},
+		{
+			name:           "UserAssignedIdentityID is specified in cloud provider config",
+			customClientID: "",
+			setupCloudProviderConfig: func(t *testing.T, config *cloud.ProviderConfig) {
+				config.UserAssignedIdentityID = "kubelet-identity-id"
+			},
+			expectToken: "token",
+			expectErr:   nil,
+		},
+		{
+			name:           "a custom client ID is specified",
+			customClientID: "custom",
+			setupCloudProviderConfig: func(t *testing.T, config *cloud.ProviderConfig) {
+				config.UserAssignedIdentityID = "kubelet-identity-id"
+			},
+			expectToken: "token",
+			expectErr:   nil,
+		},
+		{
+			name:           "service principal client secret does not contain certificate data",
+			customClientID: "",
+			setupCloudProviderConfig: func(t *testing.T, config *cloud.ProviderConfig) {
+				config.CloudName = azure.PublicCloud.Name
+				config.ClientID = "service-principal-id"
+				config.ClientSecret = "secret"
+			},
+			expectToken: "token",
+			expectErr:   nil,
+		},
+		{
+			name:           "service principal client secret contains certificate data",
+			customClientID: "",
+			setupCloudProviderConfig: func(t *testing.T, config *cloud.ProviderConfig) {
 				certData, err := testutil.GenerateCertAndKeyAsEncodedPFXData(testutil.CertTemplate{
 					CommonName:   "aad",
 					Organization: "azure",
 					Expiration:   time.Now().Add(time.Hour),
 				})
-				Expect(err).To(BeNil())
-
-				cloudProviderConfig := &cloud.ProviderConfig{
-					CloudName:    azure.PublicCloud.Name,
-					ClientID:     "service-principal-id",
-					ClientSecret: "certificate:" + certData,
-					TenantID:     testTenantID,
-				}
+				assert.NoError(t, err)
 
-				token, err := bootstrapClient.getAccessToken("", testResource, cloudProviderConfig)
-				Expect(err).To(BeNil())
-				Expect(token).To(Equal("token"))
-			})
+				config.CloudName = azure.PublicCloud.Name
+				config.ClientID = "service-principal-id"
+				config.ClientSecret = "certificate:" + certData
+			},
+			expectToken: "token",
+			expectErr:   nil,
+		},
+	}
+
+	var testResource = "resource"
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			logger, _ := zap.NewDevelopment()
+			client := &Client{
+				logger: logger,
+				extractAccessTokenFunc: func(token *adal.ServicePrincipalToken) (string, error) {
+					return tt.expectToken, tt.expectErr
```

Comment:
I think we should have each table entry define its own `extractAccessTokenFunc` and have it properly return `expectedToken` and `expectedErr` with respect to the particular test case, blindly returning `tt.expectErr` doesn't make sense I don't think